File: quicksort/quicksort.c
#include <stdio.h>
#include <stdlib.h>
#include <omp.h>
#include <stddef.h>
#include <sys/time.h>
#include <pthread.h>

int tasks;
pthread_mutex_t mtx = PTHREAD_MUTEX_INITIALIZER;

// A utility function to swap two elements
void swap(int* a, int* b)
{
    int t = *a;
    *a = *b;
    *b = t;
}
 
/* This function takes last element as pivot, places
   the pivot element at its correct position in sorted
    array, and places all smaller (smaller than pivot)
   to left of pivot and all greater elements to right
   of pivot */
int partition (int arr[], int low, int high)
{
    int pivot = arr[high];    // pivot
    int i = (low - 1);  // Index of smaller element
 
    for (int j = low; j <= high- 1; j++)
    {
        // If current element is smaller than or
        // equal to pivot
        if (arr[j] <= pivot)
        {
            i++;    // increment index of smaller element
            swap(&arr[i], &arr[j]);
        }
    }
    swap(&arr[i + 1], &arr[high]);
    return (i + 1);
}


void quickSort(int* arr, int left, int right) 
{
    pthread_mutex_lock(&mtx);
    tasks++;
    pthread_mutex_unlock(&mtx);
        int i = left, j = right;
        int tmp;
        int pivot = arr[(left + right) / 2];

        /* PARTITION PART */
        while (i <= j) {
                while (arr[i] < pivot)
                        i++;
                while (arr[j] > pivot)
                        j--;
                if (i <= j) {
                        tmp = arr[i];
                        arr[i] = arr[j];
                        arr[j] = tmp;
                        i++;
                        j--;
                }
        }
        /* RECURSION PART */
        if (left < j){ quickSort(arr, left, j);  }
        if (i< right){ quickSort(arr, i, right); }
}

void printArray(int* arr, int size)
{
    int i;
    for (i=0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}


// Driver program to test above functions
int main(int argc, char *argv[])
{
    int minNum = 1;
    int lenArr = atoi(argv[1]);
    int maxNum = lenArr;
    int* arr = (int*)malloc(lenArr * sizeof(int));

    int i;
    srand(5);  //seed
    printf("Initializing the arrays with random numbers...\n");
    for(i = 0; i < lenArr; i++) {
        arr[i] = minNum + (rand()%maxNum);
    }
    printf("Initialization finished\n");

    struct timeval start, end;
    gettimeofday(&start, NULL);

    quickSort(arr, 0, lenArr-1);

    gettimeofday(&end, NULL);

    double time = end.tv_sec + end.tv_usec/1000000.0 - (start.tv_sec + start.tv_usec/1000000.0);

    printf("tasks: %d, time: %f, throughput: %.0f\n", tasks, time, tasks/time);

    free(arr);
}

File: quicksort/quicksort.swift
import sys;
import random;
import io;

/* This function takes last element as pivot and places all smaller
 (smaller than pivot) to a new array "less" and all greater elements
 to a new array "greater". Because element of an array is future,
 no swapping can be performed. */
(int p, int o1[], int o2[]) partition(int a[], int lo, int hi) {
    pivot = a[hi];
    int m;
    int n;
    int less[];
    int greater[];
    
    //    foreach j in [lo: hi-1] {
    //      if (a[j] <= pivot) {
    //          less[m] = a[j];
    
    
    //    }
    
    for(int j = lo, m = 0, n = 0; j < hi; j = j+1, m = m+x, n = n+y){
        int x, y;
        if (a[j] <= pivot) {
            less[m] = a[j];
            x = 1;
            y = 0;
        }
        else {
            greater[n] = a[j];
            y = 1;
            x = 0;
        }
    }
    p = pivot;
    o1 = less;
    o2 = greater;
}

(int c[]) quicksort(int b[], int lo, int hi) {
    if(lo < hi) {
        pi, le, gr = partition(b, lo, hi);
        l = quicksort(le, 0, size(le) - 1);
        g = quicksort(gr, 0, size(gr) - 1);
        c = join(l, g, pi);
    }
    else {
        c = b;
    }
}


/* This function in parallel joins the array "less", pivot, array "greater"
 at their correct positions. */
(int o[]) join(int less[], int greater[], int pivot) {
    //    int A[];
    foreach i in [0:size(less)-1] {
        o[i] = less[i];
    }
    o[size(less)] = pivot;
    foreach i in [0:size(greater)-1] {
        o[i + size(less) + 1] = greater[i];
    }
    // o = A;
}

//int arr[] = [10, 7, 6, 9, 8, 1, 5, 2, 3];

int arr[];
int count = toint(argv("count"));
//int num = toint(argv("num"));
foreach i in [0:count-1] {
    //arr[i] = randint(0, num);
    arr[i] = randint(0, count);
}


start = clock() =>
int v[] = quicksort(arr, 0, size(arr)-1) =>
    time = clock() - start;
//foreach value, index in v {
//    printf("v[%i] = %s", index, value);
//}
//printf("time=%.3f, speed=%.1f", time, int2float(9 * 8)/time);
printf("time=%.3f", time);
